File: app/src/main/java/com/wafflestudio/snutt_staging/adapter/LectureListAdapter.java
package com.wafflestudio.snutt_staging.adapter;

import android.graphics.Color;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.google.common.base.Strings;
import com.wafflestudio.snutt_staging.R;
import com.wafflestudio.snutt_staging.manager.LectureManager;
import com.wafflestudio.snutt_staging.model.Lecture;

import java.util.List;

import retrofit.Callback;
import retrofit.RetrofitError;
import retrofit.client.Response;

/**
 * Created by makesource on 2017. 3. 7..
 */

public class LectureListAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    public enum VIEW_TYPE {
        Lecture(0),
        ProgressBar(1);
        private final int value;
        VIEW_TYPE(int value) {
            this.value = value;
        }
        public final int getValue() {
            return value;
        }
    }

    private static final String TAG = "LECTURE_LIST_ADAPTER" ;
    private static ClickListener clickListener;

    private List<Lecture> lectures;
    private int selectedPosition = -1;

    public LectureListAdapter(List<Lecture> lectures) {
        this.lectures = lectures;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == VIEW_TYPE.Lecture.getValue()) {
            View view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.cell_lecture, parent, false);
            // create ViewHolder
            return new LectureViewHolder(view);
        } else {
            View view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.cell_lecture_progressbar, parent, false);
            return new ProgressBarViewHolder(view);
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder viewHolder, final int position) {
        int itemType = getItemViewType(position);
        if (itemType == VIEW_TYPE.ProgressBar.getValue()) return;

        final Lecture lecture = lectures.get(position);
        final LectureViewHolder holder = (LectureViewHolder) viewHolder;
        holder.bindData(lecture);

        if (selectedPosition == position) {
            holder.layout.setBackgroundColor(Color.parseColor("#33000000"));
            if (LectureManager.getInstance().alreadyOwned(lecture)) {
                holder.add.setVisibility(View.GONE);
                holder.remove.setVisibility(View.VISIBLE);
            } else {
                holder.add.setVisibility(View.VISIBLE);
                holder.remove.setVisibility(View.GONE);
            }
        } else {
            holder.add.setVisibility(View.GONE);
            holder.remove.setVisibility(View.GONE);
            holder.layout.setBackgroundColor(Color.TRANSPARENT);
        }

        // 3개의 서로 다른 클릭을 구분하기 위해 onBindViewHolder 에서 view Id 를 비교함.
        setOnItemClickListener(new ClickListener() {
            @Override
            public void onClick(View v, final int position) {
                //TODO : (Seongowon) 배경색 바꾸기 등등 시각적 효과 넣기
                //TODO : (Seongowon) 내 강의 리스트와 비교해서 이미 있는 강의면 remove를 없으면 add버튼을 활성화
                if (v.getId() == holder.layout.getId()) {
                    Log.d(TAG, "View ID : " + v.getId());
                    Log.d(TAG, String.valueOf(position) + " item Clicked!!");
                    if (selectedPosition == position) {
                        selectedPosition = -1;
                        notifyItemChanged(position);
                        LectureManager.getInstance().setSelectedLecture(null);
                    } else {
                        notifyItemChanged(selectedPosition);
                        notifyItemChanged(position);
                        selectedPosition = position;
                        LectureManager.getInstance().setSelectedLecture(getItem(position));
                    }

                } else if (v.getId() == holder.add.getId()) {
                    Log.d(TAG, "View ID : " + v.getId());
                    Log.d(TAG, String.valueOf(position) + " add Clicked!!");
                    LectureManager.getInstance().addLecture(getItem(position), new Callback() {
                        @Override
                        public void success(Object o, Response response) {
                            notifyItemChanged(position);
                        }
                        @Override
                        public void failure(RetrofitError error) {
                        }
                    });
                } else {
                    Log.d(TAG, "View ID : " + v.getId());
                    Log.d(TAG, String.valueOf(position) + " remove Clicked!!");
                    LectureManager.getInstance().removeLecture(getItem(position), new Callback() {
                        @Override
                        public void success(Object o, Response response) {
                            notifyItemChanged(position);
                        }
                        @Override
                        public void failure(RetrofitError error) {
                        }
                    });
                }
            }
        });
    }

    @Override
    public int getItemViewType(int position) {
        Lecture item = lectures.get(position);
        return (item == null) ? VIEW_TYPE.ProgressBar.getValue() : VIEW_TYPE.Lecture.getValue();
    }

    @Override
    public int getItemCount() {
        return lectures.size();
    }

    public Lecture getItem(int position) {
        return lectures.get(position);
    }

    private static class LectureViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        protected View layout;
        protected TextView title;
        protected TextView tag;
        protected TextView classTime;
        protected TextView location;
        protected Button add;
        protected Button remove;

        private LectureViewHolder(View itemView) {
            super(itemView);
            layout = itemView;
            title = (TextView) itemView.findViewById(R.id.title);
            tag = (TextView) itemView.findViewById(R.id.tag);
            classTime = (TextView) itemView.findViewById(R.id.time);
            location = (TextView) itemView.findViewById(R.id.location);
            add = (Button) itemView.findViewById(R.id.add);
            remove = (Button) itemView.findViewById(R.id.remove);
            this.layout.setOnClickListener(this);
            this.add.setOnClickListener(this);
            this.remove.setOnClickListener(this);
        }

        private void bindData(Lecture lecture) {
            String titleText = "" ;
            titleText += lecture.getCourse_title();
            titleText += " (" + lecture.getInstructor() + " / " + String.valueOf(lecture.getCredit()) + "학점)";
            title.setText(titleText);

            String tagText = lecture.getCategory() + ", " + lecture.getDepartment() + ", "
                    + lecture.getAcademic_year();
            tag.setText(tagText);

            String classTimeText = lecture.getSimplifiedClassTime();
            if (Strings.isNullOrEmpty(classTimeText)) classTimeText = "(없음)";
            classTime.setText(classTimeText);

            String locationText = lecture.getSimplifiedLocation();
            if (Strings.isNullOrEmpty(locationText)) locationText = "(없음)";
            location.setText(locationText);
        }

        @Override
        public void onClick(View v) {
            if (clickListener != null) {
                clickListener.onClick(v, getPosition());
            }
        }
    }

    private static class ProgressBarViewHolder extends RecyclerView.ViewHolder {
        private ProgressBar progressBar;

        private ProgressBarViewHolder(View view) {
            super(view);
            progressBar = (ProgressBar) view.findViewById(R.id.progressBar);
        }
    }

    private interface ClickListener {
        public void onClick(View v, int position);
    }

    private void setOnItemClickListener(ClickListener clickListener) {
        this.clickListener = clickListener;
    }
}

File: app/src/main/java/com/wafflestudio/snutt2/adapter/LectureListAdapter.kt
package com.wafflestudio.snutt2.adapter

import android.graphics.Color
import android.text.TextUtils
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.google.common.base.Strings
import com.wafflestudio.snutt2.R
import com.wafflestudio.snutt2.SNUTTUtils.displayWidth
import com.wafflestudio.snutt2.SNUTTUtils.dp2px
import com.wafflestudio.snutt2.handler.ApiOnError
import com.wafflestudio.snutt2.lib.rx.RxBindable
import com.wafflestudio.snutt2.manager.LectureManager
import com.wafflestudio.snutt2.network.SNUTTStringUtils
import com.wafflestudio.snutt2.network.dto.core.LectureDto
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.kotlin.subscribeBy

/**
 * Created by makesource on 2017. 3. 7..
 */
class LectureListAdapter(
    private val lectures: List<LectureDto?>,
    private val lectureManager: LectureManager,
    private val apiOnError: ApiOnError,
    private val bindable: RxBindable
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    enum class VIEW_TYPE(val value: Int) {
        Lecture(0), ProgressBar(1);
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == VIEW_TYPE.Lecture.value) {
            val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.cell_lecture, parent, false)
            // create ViewHolder
            LectureViewHolder(view, lectureManager)
        } else {
            val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.cell_lecture_progressbar, parent, false)
            ProgressBarViewHolder(view)
        }
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        val itemType = getItemViewType(position)
        if (itemType == VIEW_TYPE.ProgressBar.value) return
        val lecture = lectures[position]
        val holder = viewHolder as LectureViewHolder
        val selectedPosition = lectureManager.selectedPosition
        holder.bindData(lecture!!)
        if (selectedPosition == position) {
            holder.layout.setBackgroundColor(Color.parseColor("#33000000"))
            if (lectureManager.alreadyOwned(lecture)) {
                holder.add.visibility = View.GONE
                holder.margin.visibility = View.GONE
                holder.remove.visibility = View.VISIBLE
            } else {
                holder.remove.visibility = View.GONE
                holder.margin.visibility = View.GONE
                holder.add.visibility = View.VISIBLE
            }
        } else {
            holder.add.visibility = View.GONE
            holder.remove.visibility = View.GONE
            holder.margin.visibility = View.VISIBLE
            holder.layout.setBackgroundColor(Color.TRANSPARENT)
        }

        // 3개의 서로 다른 클릭을 구분하기 위해 onBindViewHolder 에서 view Id 를 비교함.
        setOnItemClickListener(
            object : ClickListener {
                override fun onClick(v: View, position: Int) {
                    if (position >= itemCount) return
                    if (v.id == holder.layout.id) {
                        Log.d(TAG, "View ID : " + v.id)
                        Log.d(TAG, "$position item Clicked!!")
                        if (selectedPosition == position) {
                            notifyItemChanged(position)
                            lectureManager.setSelectedLecture(null)
                        } else {
                            notifyItemChanged(selectedPosition)
                            notifyItemChanged(position)
                            lectureManager.setSelectedLecture(getItem(position))
                        }
                    } else if (v.id == holder.add.id) {
                        Log.d(TAG, "View ID : " + v.id)
                        Log.d(TAG, "$position add Clicked!!")
                        lectureManager.addLecture(getItem(position)!!)
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribeBy(
                                onSuccess = {
                                    notifyItemChanged(position)
                                },
                                onError = {
                                    apiOnError(it)
                                }
                            ).let { bindable.bindDisposable(it) }
                    } else {
                        Log.d(TAG, "View ID : " + v.id)
                        Log.d(TAG, "$position remove Clicked!!")
                        lectureManager.removeLecture(getItem(position)!!)
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribeBy(
                                onSuccess = {
                                    notifyItemChanged(position)
                                },
                                onError = {
                                    apiOnError(it)
                                }
                            ).let { bindable.bindDisposable(it) }
                    }
                }
            }
        )
    }

    override fun getItemViewType(position: Int): Int {
        val item = lectures[position]
        return if (item == null) VIEW_TYPE.ProgressBar.value else VIEW_TYPE.Lecture.value
    }

    override fun getItemCount(): Int {
        return lectures.size
    }

    fun getItem(position: Int): LectureDto? {
        return lectures[position]
    }

    class LectureViewHolder constructor(var layout: View, val lectureManager: LectureManager) : RecyclerView.ViewHolder(layout), View.OnClickListener {
        protected var title: TextView
        protected var subTitle: TextView
        protected var tag: TextView
        protected var classTime: TextView
        protected var location: TextView
        var add: Button
        var remove: Button
        var margin: View
        fun bindData(lecture: LectureDto) {
            val selectedPosition = lectureManager.selectedPosition
            val titleText = lecture.course_title
            val subTitleText = "(" + lecture.instructor + " / " + java.lang.String.valueOf(
                lecture.credit
            ) + "학점)"
            title.text = titleText
            title.textScaleX = 1.0f
            subTitle.text = subTitleText
            val selected = position == selectedPosition
            val margin = 20 + 20 + 10 + if (selected) 80 else 0
            val maxWidth = (itemView.context.displayWidth - itemView.context.dp2px(margin.toFloat())).toInt()
            var subTitleWidth = Math.min(getTextViewWidth(subTitle), (maxWidth / 2).toFloat()).toInt()
            val titleWidth = Math.min(getTextViewWidth(title), (maxWidth - subTitleWidth).toFloat()).toInt()
            if (titleWidth + subTitleWidth < maxWidth) {
                subTitleWidth = maxWidth - titleWidth
            }
            title.layoutParams = LinearLayout.LayoutParams(
                titleWidth,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
            title.isSelected = selected
            title.ellipsize = if (selected) TextUtils.TruncateAt.MARQUEE else TextUtils.TruncateAt.END
            title.textScaleX = 1.0f
            subTitle.layoutParams = LinearLayout.LayoutParams(
                subTitleWidth,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
            var tagText: String? = ""
            lecture.category?.let {
                tagText += "$it, "

            }
            lecture.department?.let {
                tagText += "$it, "
            }
            tagText += lecture.academic_year
            if (Strings.isNullOrEmpty(tagText)) tagText = "(없음)"
            if (selected) {
                if (!Strings.isNullOrEmpty(lecture.remark)) {
                    tag.text = lecture.remark
                } else {
                    tag.text = tagText
                }
                tag.isSelected = true
                tag.ellipsize = TextUtils.TruncateAt.MARQUEE
            } else {
                tag.text = tagText
                tag.isSelected = false
                tag.ellipsize = TextUtils.TruncateAt.END
            }
            var classTimeText = SNUTTStringUtils.getSimplifiedClassTime(lecture)
            if (Strings.isNullOrEmpty(classTimeText)) classTimeText = "(없음)"
            classTime.text = classTimeText
            var locationText = SNUTTStringUtils.getSimplifiedLocation(lecture)
            if (Strings.isNullOrEmpty(locationText)) locationText = "(없음)"
            location.text = locationText
        }

        private fun getTextViewWidth(textView: TextView): Float {
            textView.measure(0, 0)
            return textView.measuredWidth.toFloat()
        }

        override fun onClick(v: View) {
            if (clickListener != null) {
                clickListener!!.onClick(v, position)
            }
        }

        init {
            title = itemView.findViewById<View>(R.id.title) as TextView
            subTitle = itemView.findViewById<View>(R.id.sub_title) as TextView
            tag = itemView.findViewById<View>(R.id.tag) as TextView
            classTime = itemView.findViewById<View>(R.id.time) as TextView
            location = itemView.findViewById<View>(R.id.location) as TextView
            add = itemView.findViewById<View>(R.id.add) as Button
            remove = itemView.findViewById<View>(R.id.remove) as Button
            margin = itemView.findViewById(R.id.margin)
            layout.setOnClickListener(this)
            add.setOnClickListener(this)
            remove.setOnClickListener(this)
        }
    }

    class ProgressBarViewHolder constructor(view: View) : RecyclerView.ViewHolder(view) {
        private val progressBar: ProgressBar

        init {
            progressBar = view.findViewById<View>(R.id.progressBar) as ProgressBar
        }
    }

    private interface ClickListener {
        fun onClick(v: View, position: Int)
    }

    private fun setOnItemClickListener(_clickListener: ClickListener) {
        clickListener = _clickListener
    }

    companion object {
        private const val TAG = "LECTURE_LIST_ADAPTER"
        private var clickListener: ClickListener? = null
    }
}
